zbuf/batch.go

```diff
@@ -199,40 +199,39 @@ func (r *pullerReader) Read() (*zed.Value, error) {
 	return val, nil
 }
 
-// XXX at some point the stacked scopes should not make copies of values
-// but merely refer back to the value in the wrapped batch, and we should
-// ref the wrapped batch then downstream entities will unref it, but how
-// do we carry the var frame through... protocol needs to be that any new
-// batch created by a proc needs to preserve the var frame... we don't
-// do that right now and ref counting needs to account for the dependencies.
-// procs like summarize and sort that unref their input batches merely need
-// to copy the first frame (of each batch) and the contract is that the
-// frame will not change between multiple batches within a single-EOS event.
-
 type batch struct {
-	Batch
-	vars []zed.Value
+	refs   int32
+	parent Batch
+	vals   []zed.Value
+	vars   []zed.Value
 }
 
 func NewBatch(b Batch, vals []zed.Value) Batch {
+	b.Ref()
```

Comment:
Instead of keeping the parent batch around, why not require that the invoker do a fully copy of the values that took from the parent? Seems like this would require keeping around memory that isn't be used in a lot of cases. What am I missing?

Reply:
Glad you asked.

At present we almost always do that copy, so holding onto the parent batch is pointless, wasting memory with no reduction in copying, so I'll close this.